test/jasmine/tests/transform_groupby_test.js

```diff
@@ -0,0 +1,523 @@
+            x: [1, -1, -2, 0, 1, 2, 3],
+            y: [1, 2, 3, 1, 2, 3, 1],
+
+            // everything is present:
+            transforms: [{
+                type: 'groupby',
+                groups: ['a', 'a', 'b', 'a', 'b', 'b', 'a'],
+                style: { a: {marker: {color: 'red'}}, b: {marker: {color: 'blue'}} }
+            }]
+        }];
+
+        var mockData0 = [{
+            mode: 'markers',
+            x: [1, -1, -2, 0, 1, 2, 3],
+            y: [1, 2, 3, 1, 2, 3, 1],
+
+            // groups, styles not present
+            transforms: [{
+                type: 'groupby'
+                // groups not present
+                // styles not present
+            }]
+        }];
+
+        // transform attribute with empty list
+        var mockData1 = [{
+            mode: 'markers',
+            x: [1, -1, -2, 0, 1, 2, 3],
+            y: [1, 2, 3, 1, 2, 3, 1],
+
+            // transforms is present but there are no items in it
+            transforms: [ /* list is empty */ ]
+        }];
+
+        // transform attribute with null value
+        var mockData2 = [{
+            mode: 'markers',
+            x: [1, -1, -2, 0, 1, 2, 3],
+            y: [1, 2, 3, 1, 2, 3, 1],
+            transforms: null
+        }];
+
+        // no transform is present at all
+        var mockData3 = [{
+            mode: 'markers',
+            x: [1, -1, -2, 0, 1, 2, 3],
+            y: [1, 2, 3, 1, 2, 3, 1]
+        }];
+
+        afterEach(destroyGraphDiv);
+
+        it('Plotly.plot should plot the transform traces', function(done) {
+            var data = Lib.extendDeep([], mockData);
+
+            var gd = createGraphDiv();
+
+            Plotly.plot(gd, data).then(function() {
+                expect(gd.data.length).toEqual(1);
+                expect(gd.data[0].x).toEqual([1, -1, -2, 0, 1, 2, 3]);
+                expect(gd.data[0].y).toEqual([1, 2, 3, 1, 2, 3, 1]);
+
+                expect(gd._fullData.length).toEqual(2); // two groups
+                expect(gd._fullData[0].x).toEqual([1, -1, 0, 3]);
+                expect(gd._fullData[0].y).toEqual([1, 2, 1, 1]);
+                expect(gd._fullData[1].x).toEqual([-2, 1, 2]);
+                expect(gd._fullData[1].y).toEqual([3, 2, 3]);
+
+                assertDims([4, 3]);
+
+                done();
+            });
+        });
+
+        // passes; maybe not for the good reasons (see fixme comments)
+        it('Plotly.plot should plot the transform traces', function(done) {
+            var data = Lib.extendDeep([], mockData0);
+
+            var gd = createGraphDiv();
+
+            Plotly.plot(gd, data).then(function() {
+                expect(gd.data.length).toEqual(1);
+                expect(gd.data[0].x).toEqual([1, -1, -2, 0, 1, 2, 3]);
+                expect(gd.data[0].y).toEqual([1, 2, 3, 1, 2, 3, 1]);
+
+                expect(gd._fullData.length).toEqual(0); // fixme: it passes with 0; shouldn't it be 1? (one implied group)
```

Comment:
> 1. the user doesn't specify a transforms.groups

skip over the transform

> 2. the user does specify a transforms.groups but it's null, [], undefined

same here, skip over the transform

> 3. the user specifies a transforms.groups array but its length is shorter than the length of the (shortest?) data array

The resulting extended should as many points as items in `groups`. In other words, skip over trace items that can't be paired up with a `groups` item